sitetree/tests/test_utils.py

```diff
@@ -96,13 +96,16 @@ def test_bad_string_permissions(self):
         from sitetree.toolbox import item
 
         with pytest.raises(ValueError):
-            item('root', 'url', access_by_perms='bad name')
+            # Force setup to ensure evaluation
+            item('root', 'url', access_by_perms='bad name').permissions._setup()
```

Comment:
Is there a nice way to trigger the lazy permission? For instance by doing a count() to check if it hasn't been added?

Reply:
I think that adding some other method (such as `len(item(...))`) that would trigger the `_setup()` as a side-effect (through `__len__()`) isn't better than calling a private method in this test.

Reply:
In the test we need to ensure that the existing machinery acessing `.permissions` (e.g. sitetreeapp.py:575 and in resync management command) works as expected. 

So, something like the following fixture is expected rather than `_setup()`, `any()`, `len()`, or similar:

```
        register_dynamic_trees(compose_dynamic_tree([tree('dynamic1', items=[item])]), reset_cache=True)

        with pytest.raises(ValueError) as e:
            template_render_tag('sitetree', f'sitetree_page_title from "dynamic1"', template_context(request='/'))
```

Reply:
I agree. Something that emulates real use would be better. (though it would become more of an integration test than a unittest that way 🤔 ). Your fixture gives an `django.core.exceptions.ImproperlyConfigured` in the `get_model_class()` however. 

Since I'm not very fluent in the codebase, do you have other suggestions on how to trigger the evaluasion?

Reply:
I'd rather call it a functional test %) As you can see, the test suite is comprised of different kind of test.

> Your fixture gives an django.core.exceptions.ImproperlyConfigured in the get_model_class() however.

Hm, interesting. 
Can you try something like the following and come back with a traceback?

```
    def test_bad_string_permissions(self, template_context, template_render_tag):
        from sitetree.toolbox import register_dynamic_trees, tree, item, compose_dynamic_tree

        register_dynamic_trees(compose_dynamic_tree([tree('dynamic1', items=[
            item('root', 'url', access_by_perms='bad name'),
        ])]), reset_cache=True)

        with pytest.raises(ValueError) as e:
            result = template_render_tag(
                'sitetree', f'sitetree_page_title from "dynamic1"',
                template_context(request='/'))
       
        ...

```

Reply:
Yeah I agree.

I copied your code. Here is the result:

```
==================================================================== FAILURES =====================================================================
___________________________________________________ TestPermissions.test_bad_string_permissions ___________________________________________________

self = <sitetree.tests.test_utils.TestPermissions object at 0x106bd0220>
template_context = <function template_context.<locals>.template_context_ at 0x106d1ad30>
template_render_tag = <function template_render_tag.<locals>.template_render_tag_ at 0x106d1a940>

    def test_bad_string_permissions(self, template_context, template_render_tag):
        from sitetree.toolbox import register_dynamic_trees, tree, item, compose_dynamic_tree
    
>       register_dynamic_trees(compose_dynamic_tree([tree('dynamic1', items=[
            item('root', 'url', access_by_perms='bad name'),
        ])]), reset_cache=True)

sitetree/tests/test_utils.py:98: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sitetree/utils.py:36: in tree
    tree_obj = get_tree_model()(alias=alias, title=title, **kwargs)
sitetree/utils.py:241: in get_tree_model
    return get_model_class('MODEL_TREE')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

settings_entry_name = 'MODEL_TREE'

    def get_model_class(settings_entry_name: str):
        """Returns a certain sitetree model as defined in the project settings.
    
        :param settings_entry_name:
    
        """
        app_name, model_name = get_app_n_model(settings_entry_name)
    
        try:
            model = apps_get_model(app_name, model_name)
    
        except (LookupError, ValueError):
            model = None
    
        if model is None:
>           raise ImproperlyConfigured(
                f'`SITETREE_{settings_entry_name}` refers to model `{model_name}` that has not been installed.')
E           django.core.exceptions.ImproperlyConfigured: `SITETREE_MODEL_TREE` refers to model `Model` that has not been installed.

sitetree/utils.py:233: ImproperlyConfigured
```

Reply:
Ah, yes. It was a nasty side effect of `test_import_project_sitetree_modules`.
It seems to be fixed in master now. Sorry for the inconvenience, please rebase and give it another try.

Reply:
Yes, that did the trick. I rebased and changed the tests.